### PR TITLE
[rustdoc][JSON] Use exclusively externally tagged enums in the JSON representation

### PR DESCRIPTION
**Pull request by LukeMathWalker**
See [this Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/.28De.29serialization.20speed.20of.20JSON.20docs) and [this issue](https://github.com/rust-lang/rust/issues/93667) for the relevant context.

### CODE REVIEW

**Comment by aDotInTheVoid** [comment]
This also needs to bump `FORMAT_VERSION`, as it changes the serialized representation of JSON (aside: why didn't rustbot fire on this)

**Comment by LukeMathWalker** [comment]
Milestone achieved: more than half of the test suite is green now, but many more to go 😭

File: tests/rustdoc-json/nested.rs [hunk 1]
```diff
@@ -2,30 +2,31 @@
...
     }
-    // @is "$.index[*][?(@.inner.source=='l3::L4')].kind" \"import\"
-    // @is "$.index[*][?(@.inner.source=='l3::L4')].inner.glob" false
-    // @is "$.index[*][?(@.inner.source=='l3::L4')].inner.id" $l4_id
-    // @set l4_use_id = "$.index[*][?(@.inner.source=='l3::L4')].id"
+    // @is "$.index[*][?(@.inner.import)].inner.import.glob" false
+    // @is "$.index[*][?(@.inner.import)].inner.import.source" '"l3::L4"'
+    // @is "$.index[*][?(@.inner.import)].inner.import.glob" false
+    // @is "$.index[*][?(@.inner.import)].inner.import.id" $l4_id
+    // @set l4_use_id = "$.index[*][?(@.inner.import)].id"
```
[the remark this answers is not in the document]
Comment (aDotInTheVoid) [reply tests/rustdoc-json/nested.rs:29]: Looks like https://github.com/freestrings/jsonpath/issues/91. I'd be fine with adding a `FIXME` here. CC #110406

[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply tests/rustdoc-json/nested.rs:29]: Yeah, that's precisely the issue.
I can reword most tests to work around the problem, so it's not a blocker here, but just wanted to confirm the root cause.

**Review by obi1kenobi — APPROVED** [review]
🔥🚀 Thanks, Luca!

**Review by aDotInTheVoid — CHANGES_REQUESTED** [review]
This is a heroic effort, thanks so much for doing it.

In addition to the minor things above, could you rebase this to follow the [no merge policy](https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy), and to squash all the test changes into a single commit

**Comment by aDotInTheVoid** [comment]
Also, it's worth adding `bincode` or `postcard` as a dev-dep to rustdoc-json-types, so we can test that the format is (de)serializable with non-self-describing formats, even if we don't plan to ship them in rustdoc itself.

File: tests/rustdoc-json/reexport/rename_private.rs [hunk 2]
```diff
@@ -3,11 +3,10 @@
 #![no_core]
 #![feature(no_core)]
 
-// @!has "$.index[*][?(@.kind=='inner')]"
```
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply tests/rustdoc-json/reexport/rename_private.rs:6]: I was kind of confused here, thanks for clarifying!

File: tests/rustdoc-json/reexport/glob_collision.rs [hunk 3]
```diff
@@ -3,26 +3,28 @@
 #![feature(no_core)]
 #![no_core]
 
-// @set m1 = "$.index[*][?(@.name == 'm1' && @.kind == 'module')].id"
-// @is "$.index[*][?(@.name == 'm1' && @.kind == 'module')].inner.items" []
-// @is "$.index[*][?(@.name == 'm1' && @.kind == 'module')].inner.is_stripped" true
+// @set m1 = "$.index[*][?(@.name == 'm1' && @.inner.module)].id"
+// @is "$.index[*][?(@.name == 'm1')].inner[?(@.module)].module.items" []
+// @is "$.index[*][?(@.name == 'm1')].inner[?(@.module)].module.is_stripped" true
```
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply tests/rustdoc-json/reexport/glob_collision.rs]: You're right, removed it in all cases in this file.

File: tests/rustdoc-json/enums/use_variant_foreign.rs [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 extern crate color;
 
-// @is "$.index[*][?(@.inner.name == 'Red')].kind" '"import"'
+// @has "$.index[*][?(@.inner.import)].inner.import"
```
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply tests/rustdoc-json/enums/use_variant_foreign.rs]: Yeah, I messed up the porting here. Fixed it!

**Comment by LukeMathWalker** [comment]
I've added `bincode` to the existing roundtrip tests—let me know if that suffices.
All commits have been squashed 👌🏻

**Comment by aDotInTheVoid** [comment]
r=me when CI passes

**Comment by obi1kenobi** [comment]
Thank you both for getting this in so quickly! 🙌

**Comment by aDotInTheVoid** [comment]
Don't celebrate till it gets through bors, you'll jinx it 😅

**Comment by GuillaumeGomez** [comment]
Just passing through to say: nice job everyone! :)

**Pull request merged**